neuraxle/pipeline.py

```diff
@@ -268,71 +325,131 @@ def _transform_core(self, data_container: DataContainer) -> DataContainer:
+            starting_step_data_container = step.read_checkpoint(starting_step_data_container)
 
-        return self.steps_as_tuple[new_starting_step_index:], data_container
+        return self.steps_as_tuple[new_starting_step_index:], starting_step_data_container
 
-    def _find_starting_step_index(self, data_container: DataContainer) -> int:
+    def _try_loading_pipeline_checkpoint(self, pipeline: 'Pipeline', new_starting_step_index) -> bool:
+        """
+        Load persisted pipeline if the bytecode of the steps before
+        the latest checkpoint have not changed
+
+        :param pipeline: loaded cached pipeline
+        :param new_starting_step_index:
+        :return: true if loading succeeded
+        """
+        steps_before_checkpoint_changed = False
+        for index, (step_name, step) in enumerate(self.steps_as_tuple[:new_starting_step_index + 1]):
+            source_current_step = inspect.getsource(step.__class__)
+            source_cached_step = inspect.getsource(pipeline.steps_as_tuple[index][1].__class__)
+
+            if source_current_step != source_cached_step:
+                steps_before_checkpoint_changed = True
+
+        if not steps_before_checkpoint_changed:
+            self._load_cached_pipeline_steps(new_starting_step_index, pipeline)
+            return True
+
+        return False
+
+    def _load_cached_pipeline_steps(self, new_starting_step_index, pipeline):
+        self.set_hyperparams(pipeline.get_hyperparams())
+        self.set_hyperparams_space(pipeline.get_hyperparams_space())
+        self.set_steps(
+            pipeline.steps_as_tuple[:new_starting_step_index] +
+            self.steps_as_tuple[new_starting_step_index:]
+        )
+
+    def _find_starting_step_index(self, data_container: DataContainer) -> Tuple[int, DataContainer]:
         """
         Find the index of the latest step that can be resumed
 
         :param data_container: the data container to resume
-        :return: int index latest resumable step
+        :return: index of the latest resumable step, data container at starting step
         """
-        new_data_container = copy(data_container)
+        starting_step_data_container = copy(data_container)
+        current_data_container = copy(data_container)
         index_latest_checkpoint = 0
 
         for index, (step_name, step) in enumerate(self.items()):
-            if isinstance(step, ResumableStepMixin) and step.should_resume(new_data_container):
-                data_container.set_current_ids(new_data_container.current_ids)
+            if isinstance(step, ResumableStepMixin) and step.should_resume(current_data_container):
                 index_latest_checkpoint = index
+                starting_step_data_container = copy(current_data_container)
 
             current_ids = step.hasher.hash(
-                current_ids=new_data_container.current_ids,
+                current_ids=current_data_container.current_ids,
```

Comment:
Law of Demeter is broken here by chaining two calls like `step.hasher.hash`. With a TDA solution, we could do `step.do_something()`. This is the smell that this code should be done inside the step. 

I'm not sure to understand everything in this code right here where I comment. But would the concept of a method `fake_handle_...` make sense? Could there be another way at a high level to be able to avoid faking this by rethinking the OOP design?

Reply:
Note: I'd suggest to also create another type for a faked data container to completely disable the data such that the OOP contract of that other class makes impossible looping over the data_inputs' values.

Reply:
'Law of Demeter is broken here by chaining two calls like step.hasher.hash. With a TDA solution, we could do step.do_something()' this is what I had done at first but you told me to change it lol

Reply:
the hasher code should be inside the BaseStep code  without even having a hasher object. each step would be responsible to hash the data container in a different way by overriding the hash method...